# Классификация. Деревья решений

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import seaborn as sns

sns.set()

%config InlineBackend.figure_format = 'svg'
%matplotlib inline
DATA_URL = "https://raw.githubusercontent.com/Yorko/mlcourse.ai/main/data/"

In [2]:
data= pd.read_csv(DATA_URL + "telecom_churn.csv") 

In [3]:
data.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [4]:
from sklearn.tree import DecisionTreeClassifier

In [5]:
data.drop(["State", "Voice mail plan"], axis=1, inplace=True)

In [6]:
data["International plan"] = data["International plan"].map({"Yes": 1, "No": 0})

In [7]:
data.head()

Account length  Area code  International plan  Number vmail messages  \
0             128        415                   0                     25   
1             107        415                   0                     26   
2             137        415                   0                      0   
3              84        408                   1                      0   
4              75        415                   1                      0   

   Total day minutes  Total day calls  Total day charge  Total eve minutes  \
0              265.1              110             45.07              197.4   
1              161.6              123             27.47              195.5   
2              243.4              114             41.38              121.2   
3              299.4               71             50.90               61.9   
4              166.7              113             28.34              148.3   

   Total eve calls  Total eve charge  Total night minutes  Total night calls  \
0               99             16.78                244.7                 91   
1              103             16.62                254.4                103   
2              110             10.30                162.6                104   
3               88              5.26                196.9                 89   
4              122             12.61                186.9                121   

   Total night charge  Total intl minutes  Total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   Total intl charge  Customer service calls  Churn  
0               2.70                       1  False  
1               3.70                       1  False  
2               3.29                       0  False  
3               1.78                       2  False  
4               2.73                       3  False

In [8]:
y = data["Churn"].astype("int")

In [9]:
X = data.drop("Churn", axis=1)

In [10]:
from sklearn.model_selection import cross_val_score, train_test_split

In [11]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X, y, test_size=0.3, random_state=17
)

In [12]:
first_tree = DecisionTreeClassifier(random_state=17)

In [13]:
cross_val_score(first_tree, X_train, y_train, cv=5).mean()

0.9138423504976518

In [14]:
from sklearn.neighbors import KNeighborsClassifier

In [15]:
first_knn = KNeighborsClassifier()

## настраиваем max_depth для дерева


In [16]:
from sklearn.model_selection import GridSearchCV

In [17]:
tree_params = {"max_depth": np.arange(1, 11), "max_features": [0.5, 0.7, 1]}

In [18]:
tree_grid = GridSearchCV(first_tree, tree_params, cv=5, n_jobs=1)

In [19]:
%%time
tree_grid.fit(X_train, y_train);

CPU times: user 875 ms, sys: 49.7 ms, total: 925 ms
Wall time: 1.08 s


In [20]:
tree_grid.best_score_, tree_grid.best_params_

(0.9391366681677404, {'max_depth': 6, 'max_features': 0.7})

In [21]:
from sklearn.metrics import accuracy_score
tree_valid_pred = tree_grid.predict(X_valid) 

In [22]:
accuracy_score(y_valid, tree_valid_pred)

0.936

## TASK

In [23]:
import collections
from io import StringIO

import numpy as np
import pandas as pd
import pydotplus  # pip install pydotplus
import seaborn as sns
from ipywidgets import Image
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier, export_graphviz

from matplotlib import pyplot as plt
plt.rcParams["figure.figsize"] = (10, 8)

In [24]:
# Create dataframe with dummy variables
def create_df(dic, feature_list):
    out = pd.DataFrame(dic)
    out = pd.concat([out, pd.get_dummies(out[feature_list])], axis=1)
    out.drop(feature_list, axis=1, inplace=True)
    return out


# Some feature values are present in train and absent in test and vice-versa.
def intersect_features(train, test):
    common_feat = list(set(train.keys()) & set(test.keys()))
    return train[common_feat], test[common_feat]

In [25]:
features = ["Looks", "Alcoholic_beverage", "Eloquence", "Money_spent"]

In [26]:
df_train = {}
df_train["Looks"] = [
    "handsome",
    "handsome",
    "handsome",
    "repulsive",
    "repulsive",
    "repulsive",
    "handsome",
]
df_train["Alcoholic_beverage"] = ["yes", "yes", "no", "no", "yes", "yes", "yes"]
df_train["Eloquence"] = ["high", "low", "average", "average", "low", "high", "average"]
df_train["Money_spent"] = ["lots", "little", "lots", "little", "lots", "lots", "lots"]
df_train["Will_go"] = LabelEncoder().fit_transform(["+", "-", "+", "-", "-", "+", "+"])

df_train = create_df(df_train, features)
df_train

Will_go  Looks_handsome  Looks_repulsive  Alcoholic_beverage_no  \
0        0            True            False                  False   
1        1            True            False                  False   
2        0            True            False                   True   
3        1           False             True                   True   
4        1           False             True                  False   
5        0           False             True                  False   
6        0            True            False                  False   

   Alcoholic_beverage_yes  Eloquence_average  Eloquence_high  Eloquence_low  \
0                    True              False            True          False   
1                    True              False           False           True   
2                   False               True           False          False   
3                   False               True           False          False   
4                    True              False           False           True   
5                    True              False            True          False   
6                    True               True           False          False   

   Money_spent_little  Money_spent_lots  
0               False              True  
1                True             False  
2               False              True  
3                True             False  
4               False              True  
5               False              True  
6               False              True

In [27]:
df_test = {}
df_test["Looks"] = ["handsome", "handsome", "repulsive"]
df_test["Alcoholic_beverage"] = ["no", "yes", "yes"]
df_test["Eloquence"] = ["average", "high", "average"]
df_test["Money_spent"] = ["lots", "little", "lots"]
df_test = create_df(df_test, features)
df_test

Looks_handsome  Looks_repulsive  Alcoholic_beverage_no  \
0            True            False                   True   
1            True            False                  False   
2           False             True                  False   

   Alcoholic_beverage_yes  Eloquence_average  Eloquence_high  \
0                   False               True           False   
1                    True              False            True   
2                    True               True           False   

   Money_spent_little  Money_spent_lots  
0               False              True  
1                True             False  
2               False              True

In [28]:
# Some feature values are present in train and absent in test and vice-versa.
y = df_train["Will_go"]
df_train, df_test = intersect_features(train=df_train, test=df_test)
df_train

Alcoholic_beverage_no  Looks_handsome  Alcoholic_beverage_yes  \
0                  False            True                    True   
1                  False            True                    True   
2                   True            True                   False   
3                   True           False                   False   
4                  False           False                    True   
5                  False           False                    True   
6                  False            True                    True   

   Eloquence_high  Eloquence_average  Looks_repulsive  Money_spent_little  \
0            True              False            False               False   
1           False              False            False                True   
2           False               True            False               False   
3           False               True             True                True   
4           False              False             True               False   
5            True              False             True               False   
6           False               True            False               False   

   Money_spent_lots  
0              True  
1             False  
2              True  
3             False  
4              True  
5              True  
6              True

In [29]:
df_test

Alcoholic_beverage_no  Looks_handsome  Alcoholic_beverage_yes  \
0                   True            True                   False   
1                  False            True                    True   
2                  False           False                    True   

   Eloquence_high  Eloquence_average  Looks_repulsive  Money_spent_little  \
0           False               True            False               False   
1            True              False            False                True   
2           False               True             True               False   

   Money_spent_lots  
0              True  
1             False  
2              True

1. What is the entropy S0 the initial system? By system states, we mean values of the binary feature “Will_go” - 0 or 1 - two states in total.

$$
S_0 = -\frac{3}{7} \log_2\left(\frac{3}{7}\right) - \frac{4}{7} \log_2\left(\frac{4}{7}\right) = 0.985
$$

2. Let’s split the data by the feature “Looks_handsome”. What is the entropy S1 of the left group - the one with “Looks_handsome”. What is the entropy S2 in the opposite group? What is the information gain (IG) if we consider such a split?

$$
S_1 = -\frac{1}{4} \log_2\left(\frac{1}{4}\right) - \frac{3}{4} \log_2\left(\frac{3}{4}\right) = 0.811
$$

$$
S_2 = -\frac{2}{3} \log_2\left(\frac{2}{3}\right) - \frac{1}{3} \log_2\left(\frac{1}{3}\right) = 0.918
$$

$$
IG = S_0 - \frac{4}{7} S_1 - \frac{3}{7} S_2 = 0.128
$$

Train a decision tree using sklearn on the training data. You may choose any depth for the tree.

In [30]:
dt = DecisionTreeClassifier(criterion="entropy", random_state=17)
dt.fit(df_train, y);

In [31]:
balls = [1 for i in range(9)] + [0 for i in range(11)]

In [32]:
# two groups
balls_left = [1 for i in range(8)] + [0 for i in range(5)]  # 8 blue and 5 yellow
balls_right = [1 for i in range(1)] + [0 for i in range(6)]  # 1 blue and 6 yellow

In [33]:
from math import log


def entropy(a_list):
    lst = list(a_list)
    size = len(lst)
    entropy = 0
    set_elements = len(set(lst))
    if set_elements in [0, 1]:
        return 0
    for i in set(lst):
        occ = lst.count(i)
        entropy -= occ / size * log(occ / size, 2)
    return entropy

In [34]:
print(entropy(balls))  # 9 blue и 11 yellow
print(entropy(balls_left))  # 8 blue и 5 yellow
print(entropy(balls_right))  # 1 blue и 6 yellow
print(entropy([1, 2, 3, 4, 5, 6]))  # entropy of a fair 6-sided die

0.9927744539878084
0.961236604722876
0.5916727785823275
2.584962500721156


What is the entropy of the state given by the list balls_left? - 0.961

What is the entropy of a fair dice? (where we look at a dice as a system with 6 equally probable states)? - 2.58

In [35]:
# information gain calculation
def information_gain(root, left, right):
    """ root - initial data, left and right - two partitions of initial data"""

    return (
        entropy(root)
        - 1.0 * len(left) / len(root) * entropy(left)
        - 1.0 * len(right) / len(root) * entropy(right)
    )

In [36]:
print(information_gain(balls, balls_left, balls_right))

0.16088518841412436


What is the information gain from splitting the initial dataset into balls_left and balls_right ? - 0.161

In [37]:
DATA_PATH = "https://raw.githubusercontent.com/Yorko/mlcourse.ai/main/data/"

In [38]:
data_train = pd.read_csv(DATA_PATH + "adult_train.csv", sep=";")

In [39]:
data_train.tail()

Age      Workclass  fnlwgt    Education  Education_Num  \
32556   27        Private  257302   Assoc-acdm             12   
32557   40        Private  154374      HS-grad              9   
32558   58        Private  151910      HS-grad              9   
32559   22        Private  201490      HS-grad              9   
32560   52   Self-emp-inc  287927      HS-grad              9   

            Martial_Status          Occupation Relationship    Race      Sex  \
32556   Married-civ-spouse        Tech-support         Wife   White   Female   
32557   Married-civ-spouse   Machine-op-inspct      Husband   White     Male   
32558              Widowed        Adm-clerical    Unmarried   White   Female   
32559        Never-married        Adm-clerical    Own-child   White     Male   
32560   Married-civ-spouse     Exec-managerial         Wife   White   Female   

       Capital_Gain  Capital_Loss  Hours_per_week         Country  Target  
32556             0             0              38   United-States   <=50K  
32557             0             0              40   United-States    >50K  
32558             0             0              40   United-States   <=50K  
32559             0             0              20   United-States   <=50K  
32560         15024             0              40   United-States    >50K

In [40]:
data_test = pd.read_csv(DATA_PATH + "adult_test.csv", sep=";")

In [41]:
data_test.tail()

Age      Workclass    fnlwgt   Education  Education_Num  \
16277  39        Private  215419.0   Bachelors           13.0   
16278  64            NaN  321403.0     HS-grad            9.0   
16279  38        Private  374983.0   Bachelors           13.0   
16280  44        Private   83891.0   Bachelors           13.0   
16281  35   Self-emp-inc  182148.0   Bachelors           13.0   

            Martial_Status        Occupation     Relationship  \
16277             Divorced    Prof-specialty    Not-in-family   
16278              Widowed               NaN   Other-relative   
16279   Married-civ-spouse    Prof-specialty          Husband   
16280             Divorced      Adm-clerical        Own-child   
16281   Married-civ-spouse   Exec-managerial          Husband   

                      Race      Sex  Capital_Gain  Capital_Loss  \
16277                White   Female           0.0           0.0   
16278                Black     Male           0.0           0.0   
16279                White     Male           0.0           0.0   
16280   Asian-Pac-Islander     Male        5455.0           0.0   
16281                White     Male           0.0           0.0   

       Hours_per_week         Country   Target  
16277            36.0   United-States   <=50K.  
16278            40.0   United-States   <=50K.  
16279            50.0   United-States   <=50K.  
16280            40.0   United-States   <=50K.  
16281            60.0   United-States    >50K.

In [42]:
data_test = data_test[
    (data_test["Target"] == " >50K.") | (data_test["Target"] == " <=50K.")
]

# encode target variable as integer
data_train.loc[data_train["Target"] == " <=50K", "Target"] = 0
data_train.loc[data_train["Target"] == " >50K", "Target"] = 1

data_test.loc[data_test["Target"] == " <=50K.", "Target"] = 0
data_test.loc[data_test["Target"] == " >50K.", "Target"] = 1

In [43]:
data_test.dtypes

Age                object
Workclass          object
fnlwgt            float64
Education          object
Education_Num     float64
Martial_Status     object
Occupation         object
Relationship       object
Race               object
Sex                object
Capital_Gain      float64
Capital_Loss      float64
Hours_per_week    float64
Country            object
Target             object
dtype: object

In [44]:
data_test["Age"] = data_test["Age"].astype(int)
data_test["fnlwgt"] = data_test["fnlwgt"].astype(int)
data_test["Education_Num"] = data_test["Education_Num"].astype(int)
data_test["Capital_Gain"] = data_test["Capital_Gain"].astype(int)
data_test["Capital_Loss"] = data_test["Capital_Loss"].astype(int)
data_test["Hours_per_week"] = data_test["Hours_per_week"].astype(int)

data_train["Target"] = data_train["Target"].astype(int)
data_test["Target"] = data_test["Target"].astype(int)

In [45]:
y_train = data_train.pop('Target')
y_test = data_test.pop('Target')

In [46]:
categorical_columns = [
    c for c in data_train.columns if data_train[c].dtype.name == "object"
]
numerical_columns = [
    c for c in data_train.columns if data_train[c].dtype.name != "object"
]

print("categorical_columns:", categorical_columns)
print("numerical_columns:", numerical_columns)

categorical_columns: ['Workclass', 'Education', 'Martial_Status', 'Occupation', 'Relationship', 'Race', 'Sex', 'Country']
numerical_columns: ['Age', 'fnlwgt', 'Education_Num', 'Capital_Gain', 'Capital_Loss', 'Hours_per_week']


In [47]:
for c in categorical_columns:
    data_train[c].fillna(data_train[c].mode()[0], inplace=True)
    data_test[c].fillna(data_train[c].mode()[0], inplace=True)

for c in numerical_columns:
    data_train[c].fillna(data_train[c].median(), inplace=True)
    data_test[c].fillna(data_train[c].median(), inplace=True)

In [48]:
data_train = pd.concat(
    [data_train[numerical_columns], pd.get_dummies(data_train[categorical_columns])],
    axis=1,
)

data_test = pd.concat(
    [data_test[numerical_columns], pd.get_dummies(data_test[categorical_columns])],
    axis=1,
)

In [49]:
set(data_train.columns) - set(data_test.columns)

{'Country_ Holand-Netherlands'}

In [50]:
data_test["Country_ Holand-Netherlands"] = 0

In [51]:
X_train = data_train
X_test = data_test

In [52]:
tree = DecisionTreeClassifier(max_depth=3, random_state=17)
tree.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=3, random_state=17)

In [53]:
tree_predictions = tree.predict(X_test[X_train.columns])

In [54]:
accuracy_score(y_test, tree_predictions)

0.8447884036607088

In [55]:
%%time
tree_params = {"max_depth": range(2, 11)}

locally_best_tree = GridSearchCV(
    DecisionTreeClassifier(random_state=17), tree_params, cv=5
)

locally_best_tree.fit(X_train, y_train)

CPU times: user 3.3 s, sys: 154 ms, total: 3.45 s
Wall time: 3.59 s


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=17),
             param_grid={'max_depth': range(2, 11)})

In [56]:
print("Best params:", locally_best_tree.best_params_)
print("Best cross validaton score", locally_best_tree.best_score_)

Best params: {'max_depth': 9}
Best cross validaton score 0.8566384524468356


In [57]:
tuned_tree = DecisionTreeClassifier(max_depth=9, random_state=17)
tuned_tree.fit(X_train, y_train)
tuned_tree_predictions = tuned_tree.predict(X_test[X_train.columns])
accuracy_score(y_test, tuned_tree_predictions)

0.8579939807137154